src/app/server/DefaultTermsAndConditionsProvider.cpp

```diff
@@ -0,0 +1,306 @@
+constexpr chip::TLV::Tag kAcceptedAcknowledgementsVersionTag = chip::TLV::ContextTag(3);
+constexpr uint8_t kSerializationVersion                      = 1;
+
+constexpr size_t kEstimatedTlvBufferSize = chip::TLV::EstimateStructOverhead(sizeof(uint8_t),  // SerializationVersion
+                                                                             sizeof(uint16_t), // AcceptedAcknowledgements
+                                                                             sizeof(uint16_t)  // AcceptedAcknowledgementsVersion
+                                                                             ) *
+    8 * sizeof(uint16_t); // Extra space for rollback compatibility
+} // namespace
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsStorageDelegate::Init(PersistentStorageDelegate * const inPersistentStorageDelegate)
+{
+    VerifyNotNullOrReturnInvalidArgument(inPersistentStorageDelegate);
+
+    mStorageDelegate = inPersistentStorageDelegate;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsStorageDelegate::Delete()
+{
+    VerifyNotNullOrReturnUninitialized(mStorageDelegate);
+
+    VerifyNoErrorOrReturnInternal(mStorageDelegate->SyncDeleteKeyValue(kStorageKey.KeyName()));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR chip::app::DefaultTermsAndConditionsStorageDelegate::Get(Optional<TermsAndConditions> & outTermsAndConditions)
+{
+    VerifyNotNullOrReturnUninitialized(mStorageDelegate);
+
+    uint8_t serializationVersion     = 0;
+    uint16_t acknowledgements        = 0;
+    uint16_t acknowledgementsVersion = 0;
+
+    chip::TLV::TLVReader tlvReader;
+    chip::TLV::TLVType tlvContainer;
+
+    uint8_t buffer[kEstimatedTlvBufferSize] = { 0 };
+    uint16_t bufferSize                     = sizeof(buffer);
+
+    CHIP_ERROR err = mStorageDelegate->SyncGetKeyValue(kStorageKey.KeyName(), &buffer, bufferSize);
+    if (CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND == err)
+    {
+        outTermsAndConditions.ClearValue();
+        return CHIP_NO_ERROR;
+    }
+
+    VerifyNoErrorOrReturnInternal(err);
+
+    tlvReader.Init(buffer, bufferSize);
+    VerifyNoErrorOrReturnInternal(tlvReader.Next(chip::TLV::kTLVType_Structure, chip::TLV::AnonymousTag()));
+    VerifyNoErrorOrReturnInternal(tlvReader.EnterContainer(tlvContainer));
+    VerifyNoErrorOrReturnInternal(tlvReader.Next(kSerializationVersionTag));
+    VerifyNoErrorOrReturnInternal(tlvReader.Get(serializationVersion));
+    VerifyNoErrorOrReturnInternal(tlvReader.Next(kAcceptedAcknowledgementsTag));
+    VerifyNoErrorOrReturnInternal(tlvReader.Get(acknowledgements));
+    VerifyNoErrorOrReturnInternal(tlvReader.Next(kAcceptedAcknowledgementsVersionTag));
+    VerifyNoErrorOrReturnInternal(tlvReader.Get(acknowledgementsVersion));
+    VerifyNoErrorOrReturnInternal(tlvReader.ExitContainer(tlvContainer));
+
+    // Only v1 serialization format is supported
+    VerifyOrReturnInternal(kSerializationVersion == serializationVersion);
```

Comment:
Why not tolerate v > 1? Will this clean-up as intended if this ever hits, or will the caller exit elsewhere and we will get in a messed-up place which can't go on?

Reply:
I guess that would be to handle firmware rollbacks? I was assuming that the deserialization handler would be updated when new firmware versions come out. I updated the logic to support a min-max schema to help future developers.